File: New folder/python/doublyLinkedList.py
class node:
    def __init__(self,data=None,prev=None,next=None):
        self.data=data
        self.prev=prev  
        self.next=next
    
class doubleLinked:
    def __init__(self,data): 
        self.head=node(data)
    
    #modifiers
    def insertAtTail(self,data):
        temp=self.head
        while temp.next is not None: temp=temp.next
        temp.next=node(data,temp)
    
    def insertAtHead(self,data):
        newHead=node(data,None,self.head)
        self.head.prev=newHead
        self.head=newHead
        
    
    def deleteAtTail(self):
        if self.head is None:
            return
        if self.head.next is None:
            self.head=None
            return

        temp = self.head
        while temp.next is not None:
            temp=temp.next
        temp.prev.next=None
        del temp
        return

    def deleteAtHead(self):
        if self.head.data is None:
            return
        if self.head.next is None:
            self.head=None
            return 

        temp=self.head
        self.head.next.prev=None
        self.head=self.head.next
        del temp
    
    def delete(self,key):
        it=self.head
        if it.data==key:
            self.deleteAtHead()
            return
        try:
            while it.data!=key: it=it.next
            if it.next is None:
                it.prev.next=it.next
                del it
                return
            it.next.prev=it.prev
            it.prev.next=it.next
            del it
        except:
            print("Element not found")
    
    
    #accessing
    def printList(self):
        it=self.head
        print(None,end='')
        print('<->',end='')
        while it is not None:
            print(it.data,end='')
            it=it.next
            print('<->',end='')
        print(None)
        return
        
if __name__=="__main__":
    root=doubleLinked(5)
    root.insertAtHead(3)
    root.insertAtHead(4)
    root.insertAtHead(4)
    root.insertAtTail(6)
    print("Before doing the deletions: ",end='')
    root.printList()
    root.deleteAtTail()
    root.deleteAtHead()
    root.delete(3)
    root.delete(5)
    print("After doing the deletions: ",end='')
    root.printList()

File: cpp/doublyLinkedList.cpp
#include <bits/stdc++.h>
#define ll long long
#define ld long double
#define forPrintLoop(arr,n) for(int i=0;i<n;i++) std::cout << arr[i] << ' ';
#define forInputLoop(arr,n) for(int i=0;i<n;i++) std::cin >> arr[i]
#define space ' '
#define newline '\n'
using namespace std;

class node{
    public:
    int data;
    node * next;
    node *prev;
    node(int data,node * next,node *prev):data{data},next{next},prev{prev}{}
    
};

void insertAtHead(node *&head,int data){
    if(head==NULL){
        head=new node(data,NULL,NULL);
        return;
    }
    node* newHead=new node(data,head,NULL);
    head->prev=newHead;
    head=newHead;
    return;
}

void insertAtTail(node *&head,int data){
    node *temp=head;
    if(head==NULL){
        head=new node(data,NULL,NULL);
        return;
    }
    while(temp->next!=NULL){
        temp=temp->next;
    }
    temp->next = new node(data,NULL,temp);
    return;
}
void printList(node *head){
    cout << "NULL" << "<->";
    while(head!=NULL){
        cout << head->data << "<->";
        head=head->next;
    }
    cout << "NULL" << newline;
    return;
}

void deleteHead(node *&head){
    if(head==NULL){
        cout << "Empty list -> cannot pop" << newline;
        return;
    }
    if(head->next==NULL){
        delete head;
        head=NULL;
    }

    auto temp=head;
    head->next->prev=NULL;
    head=head->next;
    delete temp;

}

void deleteAtTail(node *&head){
    if(head==NULL){
        cout<<"Empty list -> cannot pop" << newline;
        return;
    }
    if(head->next==NULL){
        delete head;
        head=NULL;
        return;
    }
    auto toDel=head;
    while(toDel->next->next){
        toDel=toDel->next;
    }
    delete toDel->next;
    toDel->next=NULL;
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    node * head=NULL;
    insertAtHead(head,5);
    insertAtHead(head,3);
    cout << "After inserting 3,5 at head" << newline;
    printList(head);
    cout << "After inserting 3,4 at tail" << newline;
    insertAtTail(head,3);
    insertAtTail(head,4);
    printList(head);
    cout << "After deleting at head 2 times" << newline;
    deleteHead(head);
    deleteHead(head);
    printList(head);
    cout << "After deleting at tail 1 time" << newline;
    deleteAtTail(head);
    printList(head);
    return 0;
}